### ISSUE TITLE
Dimensions of output and target?

### ISSUE THREAD

**Issue by sarmientoj24** [mlyg/mixed-focal-loss#18]
Can it be that the output is of the form batch x num_classes x height x width while the target is batch x height x width?

**Comment by mlyg** [comment]
Hi sarmientoj24,

You will need both the output and target to be in the form: batch x num_classes x height x width. The two tensors (the output tensor and target tensor) need to be the same shape otherwise you will not be able to calculate the loss correctly. 

Best,
Michael

**Comment by sarmientoj24** [comment]
@mlyg Thank you for the quick response.

Is this correct for:
- Pytorch
- multi-class approach



```
# y_pred is in batch x num_classes x height x width
# y_true is in batch x height x width

class AsymmetricFocalTverskyLoss(BaseLoss):
    """This is the implementation for binary segmentation.
    Parameters
    ----------
    delta : float, optional
        controls weight given to false positive and false negatives, by default 0.7
    gamma : float, optional
        focal parameter controls degree of down-weighting of easy examples, by default 0.75
    smooth : float, optional
        smooithing constant to prevent division by 0 errors, by default 0.000001
    epsilon : float, optional
        clip values to prevent division by zero error
    """
    def __init__(self, num_classes, delta=0.7, gamma=0.75, epsilon=1e-07):
        self.num_classes = num_classes
        self.delta = delta
        self.gamma = gamma
        self.epsilon = epsilon

    def forward(self, y_pred, y_true):
        # Convert y_true from N x H x W to N x C x H x W of binary class 0-1
        y_true = make_one_hot(y_true.unsqueeze(1), self.num_classes)

        # Use Softmax on y_pred
        y_pred = nn.functional.softmax(y_pred, dim=1)

        # Clip values to prevent division by zero error
        y_pred = torch.clamp(y_pred, self.epsilon, 1. - self.epsilon)
        axis = identify_axis(y_true.size())

        # Calculate true positives (tp), false negatives (fn) and false positives (fp)     
        tp = torch.sum(y_true * y_pred, axis=axis)
        fn = torch.sum(y_true * (1-y_pred), axis=axis)
        fp = torch.sum((1-y_true) * y_pred, axis=axis)
        dice_class = (tp + self.epsilon)/(tp + self.delta*fn + (1-self.delta)*fp + self.epsilon)

        losses = []

        for i in range(self.num_classes):
            if i == 0:
                loss = (1-dice_class[:,i]) 

            else:
                loss = (1-dice_class[:,i]) * torch.pow(1-dice_class[:,i], -self.gamma)
            losses.append(loss)

        loss = torch.mean(torch.stack(losses, axis=-1))
        return loss
```

**Comment by sarmientoj24** [comment]
What is the ideal dtype for the `y_pred` and `y_true`? `torch.float32` and `torch.int64`?

**Comment by sarmientoj24** [comment]
Also, I am curious how these are calculated
```
        tp = torch.sum(y_true * y_pred, axis=axis)
        fn = torch.sum(y_true * (1-y_pred), axis=axis)
        fp = torch.sum((1-y_true) * y_pred, axis=axis)
```

if `y_pred` is not binary.
